lib/grandpa/grandpa.go

```diff
@@ -321,69 +319,32 @@ func (s *Service) initiateRound() error {
 
 	// make sure no votes can be validated while we are incrementing rounds
 	s.roundLock.Lock()
+	defer s.roundLock.Unlock()
+
 	s.state.round++
 	logger.Debugf("incrementing grandpa round, next round will be %d", s.state.round)
 	s.prevotes = new(sync.Map)
 	s.precommits = new(sync.Map)
 	s.pvEquivocations = make(map[ed25519.PublicKeyBytes][]*SignedVote)
 	s.pcEquivocations = make(map[ed25519.PublicKeyBytes][]*SignedVote)
-	s.roundLock.Unlock()
-
-	best, err := s.blockState.BestBlockHeader()
-	if err != nil {
-		return err
-	}
 
-	if best.Number > 0 {
-		return nil
-	}
-
-	// don't begin grandpa until we are at block 1
-	s.waitForFirstBlock()
 	return nil
 }
 
 // initiate initates the grandpa service to begin voting in sequential rounds
 func (s *Service) initiate() error {
-	for {
-		err := s.initiateRound()
-		if err != nil {
-			logger.Warnf("failed to initiate round for round %d: %s", s.state.round, err)
-			return err
-		}
-
-		err = s.playGrandpaRound()
-		if errors.Is(err, ErrServicePaused) {
-			logger.Info("service paused")
-			// wait for service to un-pause
-			<-s.resumed
-			err = s.initiate()
-		}
-
-		if err != nil {
-			logger.Warnf("failed to play grandpa round: %s", err)
-			continue
-		}
-
-		if s.ctx.Err() != nil {
-			return errors.New("context cancelled")
-		}
+	finalizationHandler := newFinalizationHandler(s)
+	errsCh, err := finalizationHandler.Start()
+	if err != nil {
+		return fmt.Errorf("starting finalisation handler: %w", err)
 	}
-}
-
-func (s *Service) waitForFirstBlock() {
-	ch := s.blockState.GetImportedBlockNotifierChannel()
-	defer s.blockState.FreeImportedBlockNotifierChannel(ch)
 
-	// loop until block 1
 	for {
 		select {
-		case block := <-ch:
-			if block != nil && block.Header.Number > 0 {
-				return
-			}
 		case <-s.ctx.Done():
-			return
+			return finalizationHandler.Stop()
```

Comment:
If the ctx gets canceled, shouldn't we also collect errors from `errsCh` (even if unchecked, to at least drain the channel and let the other goroutine exit)?